ldap/servers/plugins/alias_entries/alias-entries.c

```
@@ -0,0 +1,123 @@
+        }
+    }
+    slapi_free_search_results_internal(pb);
+    slapi_pblock_destroy(pb);
+
+    return entry_sdn;
+}
+
+int
+alias_entry_init(Slapi_PBlock *pb)
+{
+    if (slapi_pblock_set(pb, SLAPI_PLUGIN_VERSION, SLAPI_PLUGIN_VERSION_03) != 0 ||
+        slapi_pblock_set(pb, SLAPI_PLUGIN_DESCRIPTION, (void *)&srchpdesc) != 0 ||
+        slapi_pblock_set(pb, SLAPI_PLUGIN_PRE_SEARCH_FN, (void *) alias_entry_srch) != 0  ||
+        slapi_pblock_get(pb, SLAPI_PLUGIN_IDENTITY, &plugin_id) != 0)
+    {
+        slapi_log_error(SLAPI_LOG_ERR, PLUGINNAME,
+                        "alias_entry_init: Error registering the plug-in.\n");
+        return -1;
+    }
+
+    return 0;
+}
+
+int
+alias_entry_srch(Slapi_PBlock *pb)
+{
+    Slapi_DN *search_target = NULL;
+    Slapi_DN *dn1 = NULL;
+    Slapi_DN *dn2 = NULL;
+    char *str_filter = NULL;
+    int scope = 0;
+    int deref_enabled = 0;
+    size_t i = 0;
+
+    /* deref must be enabled */
+    if (slapi_pblock_get(pb, SLAPI_SEARCH_DEREF, &deref_enabled) != 0) {
+        return 0;
+    }
+    if (deref_enabled != LDAP_DEREF_FINDING && deref_enabled != LDAP_DEREF_ALWAYS) {
+        return 0;
+    }
+
+    /* Only base search is supported */
+    if (slapi_pblock_get(pb, SLAPI_SEARCH_SCOPE, &scope) != 0 || scope != LDAP_SCOPE_BASE) {
+        return 0;
+    }
+
+    /* Skip reserved operations */
+    if (slapi_op_reserved(pb)) {
+        return 0;
+    }
+
+    /* Base dn should be provided */
+    if (slapi_pblock_get(pb, SLAPI_SEARCH_TARGET_SDN, &search_target) != 0 || search_target == NULL) {
+        return 0;
+    }
+
+    /* Skip our own requests */
+    if (slapi_pblock_get(pb, SLAPI_SEARCH_STRFILTER, &str_filter) != 0 ||
+        strcmp(str_filter, ALIASFILTER) == 0)
+    {
+        return 0;
+    }
+
+    /* Follow the alias chain */
+    dn2 = search_target;
+    do {
+        dn1 = dn2;
+        dn2 = alias_get_next_dn(dn1);
+        if (i > 0 && dn2 != NULL) {
+            slapi_sdn_free(&dn1);
```

I have a doubt here.

If the function alias_get_next_dn returns NULL, dn1 is not freed. And alias_get_next_dn can return NULL if (slapi_search_internal_pb(pb) != 0).

---

So I wonder if we should change it to something like that:

```
    if (i > 0 && dn1 != search_target) {
        slapi_sdn_free(&dn1);
    }
```

So dn1 is always freed unless it is the original search_target.

---

No the code is good: We cannot not free dn1 as it is used afterward and stored in the search pblock.
  and checking for search_target is redundant with i>0 test ( because dn1 ==  search_target <==> i == 0 )
      (I agree that checking for dn1 != search_target instead of i > 0 would be clearer but code is OK. )
  dn1 is the current dereferenced target (the loop handles the case of an alias pointing on another alias case)
  so if there is a next item (i.e dn2), we should free the previous dn1 before overwriting it in next iteration.
  And the loop aborted case (because there are too many hops) is fine too.
    dn1 has been freed if not search_target by the (i > 0 && dn2 != NULL) and dn2 is freed 
     in the  dn2 == NULL ...  else code

---

My main concern is that the  `slapi_search_internal_pb(pb)` call from `alias_get_next_dn` function  may fail (for any unexpected reason).
The way `alias_get_next_dn` function is written, it will return NULL.

If it'll happen in the middle of the do-while loop here, we won't free `dn1` and it won't be `search_target` either.
So what will happen next - we'll assign `dn1` to `SLAPI_SEARCH_TARGET_SDN` and, unless I miss something, it won't be right.

But yeah, my solution is not complete for this case...

---

True, if the internal search failed (e.g. nonexistent DN in aliasedObjectName) then the parent in the chain would be returned instead.  So this can be hardened for a such a case.  

Do we log an warning?  
Do we return an error to the client?
Do we just return the parent?
Or return no result at all (err=0 nentries=0)?

Probably should just return an error to client: operations error?  Unwilling to perform?

---

My vote is for the operations error. I think we should notify the user that something is wrong and point out that dereference is broken in his requested scope.

If the admin decided to use the plugin - probably there's a good reason for that and a warning, parent's return or just no results is not enough in my opinion.

@mreynolds389 @tbordaz @progier389 thoughts? :)

---

The RFC does not provide any requirement for such a scenario.  I'm torn between not adjusting the search target at all, or returning an error (err=1 works for me).

---

We can argue about it, but IMHO its depends of the failure cause.
  if it is NOSUCHOBJECT the current code behavior is not so bad: returning the last entry that we were able to dereference. Anyway by looking at its objectclass, the client application should easily determine that it is an alias and that something is wrong. (returning the original entry is possible too)
  If it is another error, I would return the internal search error code and a result text like: "'%s' while dereferencing alias %s" internal_search_result_text, dn1

---

Ok, since this a preop plugin I can not actually cause the operation to fail, but I can attach a result text.  I also can not get python-ldap to detect this result text in the CI, but I can confirm with ldapsearch that it is there.